src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,251 @@
+     * @param rotation rhe rotational velocity. [rad/s]
+     */
+    public void holonomicDrive(double forward, double strafe, double rotation) {
+        if (rotation == 0 &&
+                rotationDelay.update(Math.abs(thetaController.getGoal().position - Robot.getAngle()
+                                .getRadians()) < Constants.SwerveDrive.ALLOWABLE_THETA_ERROR,
+                        Constants.SwerveDrive.ROTATION_DELAY)
+        ) {
+            rotation = thetaController.calculate(Robot.getAngle().getRadians());
+        } else {
+            thetaController.reset(Robot.getAngle().getRadians());
+        }
+        ChassisSpeeds speeds = fieldOriented ?
+                ChassisSpeeds.fromFieldRelativeSpeeds(forward, strafe, rotation, Robot.getAngle()) :
+                new ChassisSpeeds(forward, strafe, rotation);
+
+        setStates(kinematics.toSwerveModuleStates(speeds));
+    }
+
+    /**
+     * Rotates the robot to a desired angle.
+     *
+     * @param desiredAngle the desired angle of the robot.
+     */
+    public void holonomicDriveKeepSetpoint(double forward, double strafe, Rotation2d desiredAngle) {
+        thetaController.setGoal(desiredAngle.getRadians());
+        double output = thetaController.calculate(Robot.getAngle().getRadians());
+
+        setStates(kinematics.toSwerveModuleStates(new ChassisSpeeds(forward, strafe, output)));
+    }
+
+    /**
+     * Gets te states of every module.
+     *
+     * @return the states of every module.
+     */
+    public SwerveModuleState[] getStates() {
+        SwerveModuleState[] swerveModuleState = new SwerveModuleState[modules.length];
+        for (SwerveModule module : modules) {
+            swerveModuleState[module.getWheel()] = module.getState();
+        }
+        return swerveModuleState;
+    }
+
+    /**
+     * Sets the state of the modules.
+     *
+     * @param states the states of the modules.
+     */
+    public void setStates(SwerveModuleState[] states) {
+        SwerveDriveKinematics.desaturateWheelSpeeds(states, Constants.SwerveDrive.VELOCITY_MULTIPLIER);
+        for (SwerveModule module : modules) {
+            SwerveModuleState state = states[module.getWheel()];
+            state = SwerveModuleState.optimize(state, module.getAngle());
+            module.setState(state);
+        }
+    }
+
+    /**
+     * Gets the chassis speeds of the entire robot.
+     *
+     * @return the speed of the robot in each axis.
+     */
+    public ChassisSpeeds getChassisSpeeds() {
+        ChassisSpeeds chassisSpeeds = kinematics.toChassisSpeeds(getStates());
+        if (fieldOriented) {
+            chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(
+                    chassisSpeeds.vxMetersPerSecond,
+                    chassisSpeeds.vyMetersPerSecond,
+                    chassisSpeeds.omegaRadiansPerSecond,
+                    Robot.getAngle()
+            );
+        }
+        return chassisSpeeds;
```

Comment:
Another option to reduce indentation:
```suggestion
        if (!fieldOriented) {
            return chassisSpeeds;
        }
        chassisSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(
                chassisSpeeds.vxMetersPerSecond,
                chassisSpeeds.vyMetersPerSecond,
                chassisSpeeds.omegaRadiansPerSecond,
                Robot.getAngle()
        );
        return chassisSpeeds;
    }
```